python/ray/air/session.py

```diff
@@ -0,0 +1,53 @@
+import abc
+import logging
+from typing import Optional, Dict
+
+from ray.air.checkpoint import Checkpoint
+
+logger = logging.getLogger(__name__)
+
+
+class Session(abc.ABC):
+    """The canonical session interface that both Tune and Train session implements.
+
+    User can interact with this interface to get session information,
+    as well as reporting metrics and saving checkpoint.
+    """
+
+    @abc.abstractmethod
+    def report(self, metrics: Dict, checkpoint: Optional[Checkpoint] = None) -> None:
+        """Report metrics and optionally save checkpoint.
+
+        The checkpoint is guaranteed to be saved onto Driver node.
+        Each invocation of this method will automatically increment the underlying
+        epoch number.
+        This API is supposed to replace the legacy ``tune.report``,
+        ``with tune.checkpoint_dir``, ``train.report`` and ``train.save_checkpoint``.
+        Please avoid mixing them together.
+        """
+
+        raise NotImplementedError
+
+    # @abc.abstractmethod
```

Comment:
This is useful to enable cloud storage story for distributed training, which is currently missing in Ray Air.
With this, user can then do (inside their huggingface worker logic for example)
```
session = get_session()
if session.rank() == 0:
  checkpoint.to_uri(session.remote_checkpoint_dir())
  session.report({"foo": bar}, checkpoint=checkpoint)  // which only sends the URI string for book keeping purposes. 
```

Reply:
We decide to just use SyncConfig and not burden user with these conversions.